packages/ember-data/lib/system/store.js

```diff
@@ -2058,6 +2058,7 @@ function deserializeRecordId(store, data, key, relationship, id) {
 
   //If record objects were given to push directly, uncommon, not sure whether we should actually support
   if (id instanceof Model) {
+    Ember.deprecate("Passing a record object to push directly into the store has been deprecated. Pass a number or string as a value to represent the record.");
```

Comment:
Can you make the deprecation message a bit more descriptive to the user. Think about if you were to see it pop up in the console. Something with more context like:
You tried pushing a record 'recordType' with id 'id'  and passed a DS.Model as a value for the relationship 'relationshipKey' etc.